prow/crier/controller_test.go

```diff
@@ -0,0 +1,268 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/client/clientset/versioned/fake"
+	prowLister "k8s.io/test-infra/prow/client/listers/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+)
+
+const (
+	testTimeout    = time.Second
+	controllerName = "CrierTest"
+	reporterName   = "fakeReporter"
+)
+
+// Fake Reporter
+// Sets: Which jobs should be reported
+// Asserts: Which jobs are actually reported
+type fakeReporter struct {
+	reported         []string
+	shouldReportFunc func(pj *prowv1.ProwJob) bool
+}
+
+func (f *fakeReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	f.reported = append(f.reported, pj.Spec.Job)
+	return []*prowv1.ProwJob{pj}, nil
+}
+
+func (f *fakeReporter) GetName() string {
+	return reporterName
+}
+
+func (f *fakeReporter) ShouldReport(pj *prowv1.ProwJob) bool {
+	return f.shouldReportFunc(pj)
+}
+
+// Fake Informer
+// Sets: The Prow Job Test Cases
+type fakeInformer struct {
+	jobs map[string]*prowv1.ProwJob
+}
+
+func (f fakeInformer) Get(name string) (*prowv1.ProwJob, error) {
+	pj, found := f.jobs[name]
+	if !found {
+		var s schema.GroupResource
+		return nil, errors.NewNotFound(s, "Can't Find ProwJob")
+	}
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) prowLister.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() prowLister.ProwJobLister {
+	return f
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) AddEventHandler(handler cache.ResourceEventHandler)                        {}
+func (fakeInformer) Run(stopCh <-chan struct{})                                                {}
+func (fakeInformer) AddEventHandlerWithResyncPeriod(cache.ResourceEventHandler, time.Duration) {}
+func (fakeInformer) GetStore() cache.Store                                                     { return nil }
+func (fakeInformer) GetController() cache.Controller                                           { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                           { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                                 { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                                 { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*prowv1.ProwJob, err error)          { return nil, nil }
+
+func TestController_Run(t *testing.T) {
+	tests := []struct {
+		name         string
+		jobsOnQueue  []string
+		knownJobs    map[string]*prowv1.ProwJob
+		shouldReport func(*prowv1.ProwJob) bool
+
+		expectReports []string
+		expectPatch   bool
+	}{
+		{
+			name:        "reports/patches known job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
```

Comment:
Given that there are no examples of multiple known jobs, I'd be inclined to make this `knownJob prowv1.ProwJob`.

Given that 90% of stuff is the same, I'd be inclined to do something like:
```golang
tests := []struct{
  name string
  knownJob func(*prowv1.ProwJob)
}{
  {
    name: "report/patches known job",
  },
   {
    name: "doesn't report non-existant job",
    knownJob: func(pj *prowv1.ProwJob) {
      var newPJ prowv1.ProwJob
      *pj = newPJ
    },
  {
    name: "previous isn't re-reported",
    knownJob: func(pj *prowv1.ProwJob) {
      pj.Status.PrevReportStates = map[string]prowv1.ProwJobState{
							reporterName: prowv1.TriggeredState,
      }
    }
  },
}

for _, tc := range tests {
  var pj prowv1.ProwJob
  pj.Job = "foo"
  pj.Status = prowv1.TriggeredState
  if tc.knownJob != nil {
    tc.knownJob(pj)
  }
}
```